ethereum/src/ERC20Manager.sol

```diff
@@ -530,74 +530,79 @@ contract ERC20Manager is
             }
 
+            // we use offset `OFFSET4 = DISCRIMINANT_SIZE` to skip `uint8 discriminant`
             bytes32 vftManager;
             assembly ("memory-safe") {
-                vftManager := calldataload(payload.offset)
+                vftManager := calldataload(add(payload.offset, OFFSET4))
             }
 
-            _knownVftManagers.add(vftManager);
+            _vftManagers.add(vftManager);
 
             emit VftManagerAdded(vftManager);
+
+            return true;
         }
 
-        if (discriminant == uint256(SupplyType.Ethereum)) {
+        if (discriminant == uint256(TokenType.Ethereum)) {
             if (!(payload.length == REGISTER_ETHEREUM_TOKEN_MESSAGE_SIZE)) {
                 return false;
             }
 
-            bytes32 tokenName;
-            bytes32 tokenSymbol;
-            uint8 tokenDecimals;
-
             // we use offset `OFFSET4 = DISCRIMINANT_SIZE` to skip `uint8 discriminant`
-            // we use offset `OFFSET5 = DISCRIMINANT_SIZE + TOKEN_NAME_SIZE` to skip `uint8 discriminant` and `bytes32 tokenName`
-            // we use offset `OFFSET6 = DISCRIMINANT_SIZE + TOKEN_NAME_SIZE + TOKEN_SYMBOL_SIZE` to skip `uint8 discriminant`, `bytes32 tokenName` and `bytes32 tokenSymbol`
+            address token;
             assembly ("memory-safe") {
-                tokenName := calldataload(add(payload.offset, OFFSET4))
-                tokenSymbol := calldataload(add(payload.offset, OFFSET5))
-                tokenDecimals := calldataload(add(payload.offset, OFFSET6))
-                // `TOKEN_DECIMALS_BIT_SHIFT` right bit shift is required to remove extra bits since `calldataload` returns `uint256`
-                tokenDecimals := shr(TOKEN_DECIMALS_BIT_SHIFT, calldataload(add(payload.offset, OFFSET6)))
+                // `TOKEN2_BIT_SHIFT` right bit shift is required to remove extra bits since `calldataload` returns `uint256`
+                token := shr(TOKEN2_BIT_SHIFT, calldataload(add(payload.offset, OFFSET4)))
             }
 
-            uint8 tokenNameLength = uint8(tokenName[0]);
-            if (!(tokenNameLength >= 1 && tokenNameLength <= 31)) {
-                return false;
-            }
+            _tokens.push(token);
+            _tokenType[token] = TokenType.Ethereum;
 
-            uint8 tokenSymbolLength = uint8(tokenSymbol[0]);
-            if (!(tokenSymbolLength >= 1 && tokenSymbolLength <= 31)) {
-                return false;
-            }
+            emit EthereumTokenRegistered(token);
 
-            string memory tokenNameStr = LibString.unpackOne(tokenName);
-            string memory tokenSymbolStr = LibString.unpackOne(tokenSymbol);
+            return true;
+        }
 
-            ERC20GearSupply token = new ERC20GearSupply(address(this), tokenNameStr, tokenSymbolStr, tokenDecimals);
+        // `discriminant == uint256(TokenType.Gear)` is guaranteed by previous checks
+        if (!(payload.length == REGISTER_GEAR_TOKEN_MESSAGE_SIZE)) {
```

Comment:
It's actually creation, rather than just addition 
```suggestion
        if (!(payload.length == CREATE_GEAR_TOKEN_MESSAGE_SIZE)) {
```

Reply:
what's wrong with "registration"? it's still registration with the deployment on ERC20Manager side.